### PR TITLE
Feat/opentelemetry

### LINKED ISSUE #3213: Support for Prometheus metrics export (e.g. OpenTelemetry or prometheus_client)
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Support for exporting metrics to a format compatible with Prometheus (e.g. OpenTelemetry).

Currently, inference projects such as VLLM & SGLang support exporting metrics via /metrics endpoint. For training, particularly RL, it would be very helpful to have the training metrics from Axolotl also available for Prometheus/Grafana so disconnected environments or sensitive customers aren't reliant on public sites.

## Possible Approach 1: OpenTelemetry (Supports more apps, but possibly complex)
- Implement OTEL MeterProvider + instruments to directly collect time-series data from Axolotl training and create a /metrics endpoint that exposes these metric for Prometheus scraping.
(https://betterstack.com/community/guides/observability/otel-metrics-python/)
(https://opentelemetry.io/docs/specs/otel/metrics/api/)

Possible configuration items (via config or ENVAR):
```yaml
use_otel_metrics: 
otel_metrics_host: 
otel_metrics_port: 
```

## Possible Approach 2: Prometheus Python Client
- Implement prometheus client_python and loss (and other metrics) to a /metrics endpoint
(https://prometheus.github.io/client_python/)
- Example of VLLM observability implementation with client_python (prometheus_client): https://github.com/vllm-project/vllm/blob/main/vllm/engine/metrics.py
- VLLM Observability CLI reference: https://docs.vllm.ai/en/latest/cli/serve.html#observabilityconfig

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by ved1beta**
## Motivation and Context

#3213 

## How has this been tested?
image
logs are visible on 

`http://localhost:8000/metrics` after the run

added tests

## Summary by CodeRabbit

* **New Features**
  * Added OpenTelemetry metrics collection to export training metrics via Prometheus for real-time monitoring.
  * Added configuration options to enable metrics collection and specify the metrics server host and port.
  * Added example configuration demonstrating QLoRA training with OpenTelemetry metrics enabled.

* **Chores**
  * Added OpenTelemetry and Prometheus dependencies.

### CODE REVIEW

File: src/axolotl/utils/callbacks/opentelemetry.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            self.eval_loss_gauge.set(logs["eval_loss"])
+
+        if "learning_rate" in logs:
+            self.learning_rate_gauge.set(logs["learning_rate"])
+
+        if "epoch" in logs:
+            self.epoch_gauge.set(logs["epoch"])
+
+        if "grad_norm" in logs:
+            self.grad_norm_gauge.set(logs["grad_norm"])
```
Comment (NanoCode012) [inline src/axolotl/utils/callbacks/opentelemetry.py:174]: Should also track memory usage logs

**Comment by ed-pai** [comment]
Just some positive feedback... We were able to test out this feature and it was exactly what we were looking for. Easy to setup and provided the training metrics we needed for our Prometheus/Grafana stack. Great work!
